server/src/main/java/org/elasticsearch/repositories/RepositoryOperation.java

```diff
@@ -8,18 +8,69 @@
  */
 package org.elasticsearch.repositories;
 
+import org.elasticsearch.cluster.DiffableUtils;
+import org.elasticsearch.cluster.metadata.ProjectId;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.core.FixForMultiProject;
+
+import java.io.IOException;
+
 /**
  * Coordinates of an operation that modifies a repository, assuming that repository at a specific generation.
  */
 public interface RepositoryOperation {
 
+    /**
+     * Project for which repository belongs to.
+     */
+    @FixForMultiProject(description = "default implementation is temporary")
+    default ProjectId projectId() {
+        return ProjectId.DEFAULT;
+    }
+
     /**
      * Name of the repository affected.
      */
     String repository();
 
+    default ProjectRepo projectRepo() {
+        return new ProjectRepo(projectId(), repository());
+    }
+
     /**
      * The repository state id at the time the operation began.
      */
     long repositoryStateId();
+
+    record ProjectRepo(ProjectId projectId, String repoName) implements Writeable {
+
+        public ProjectRepo(StreamInput in) throws IOException {
+            this(ProjectId.readFrom(in), in.readString());
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            projectId.writeTo(out);
+            out.writeString(repoName);
+        }
+    }
+
+    DiffableUtils.KeySerializer<ProjectRepo> PROJECT_REPO_SERIALIZER = new DiffableUtils.KeySerializer<>() {
+        @Override
+        public void writeKey(ProjectRepo key, StreamOutput out) throws IOException {
+            key.writeTo(out);
+        }
+
+        @Override
+        public ProjectRepo readKey(StreamInput in) throws IOException {
+            return new ProjectRepo(in);
+        }
+    };
+
+    @Deprecated(forRemoval = true)
+    static ProjectRepo defaultProjectRepo(String repoName) {
+        return new ProjectRepo(ProjectId.DEFAULT, repoName);
+    }
```

Comment:
Should we add `@FixForMultiProject`? Also, wouldn't it make more sense to use `Metadata.DEFAULT_PROJECT_ID` here as this is a temporary usage?

Reply:
I deleted this method as part of the changes to add an extra ProjectId argument. See also https://github.com/elastic/elasticsearch/pull/125470#discussion_r2011193579